File: service/AuthService.java
package com.investment.tracker.service;

import com.investment.tracker.dto.LoginRequest;
import com.investment.tracker.dto.LoginResponse;
import com.investment.tracker.dto.RegisterRequest;
import com.investment.tracker.exception.InvalidCredentialsException;
import com.investment.tracker.model.Role;
import com.investment.tracker.model.User;
import com.investment.tracker.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

/**
 * Service for handling authentication operations.
 */
@Service
@RequiredArgsConstructor
@Slf4j
public class AuthService {

    private final AuthenticationManager authenticationManager;
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    /**
     * Authenticate user with username and password.
     * @param loginRequest the login credentials
     * @return LoginResponse containing user information
     * @throws InvalidCredentialsException if credentials are invalid
     */
    public LoginResponse login(LoginRequest loginRequest) {
        try {
            // Authenticate user
            Authentication authentication = authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(
                            loginRequest.getUsername(),
                            loginRequest.getPassword()
                    )
            );

            // Set authentication in security context
            SecurityContextHolder.getContext().setAuthentication(authentication);

            // Retrieve user details
            User user = userRepository.findByUsername(loginRequest.getUsername())
                    .orElseThrow(() -> new InvalidCredentialsException(
                            "User not found: " + loginRequest.getUsername()));

            log.info("User '{}' logged in successfully with role {}", user.getUsername(), user.getRole());

            // Build and return response
            return LoginResponse.builder()
                    .id(user.getId())
                    .username(user.getUsername())
                    .email(user.getEmail())
                    .role(user.getRole())
                    .message("Login successful")
                    .build();

        } catch (BadCredentialsException e) {
            log.warn("Failed login attempt for username: {}", loginRequest.getUsername());
            throw new InvalidCredentialsException("Invalid username or password");
        }
    }

    /**
     * Register a new user.
     * @param registerRequest the registration details
     * @return LoginResponse containing new user information
     * @throws IllegalArgumentException if username or email already exists
     */
    public LoginResponse register(RegisterRequest registerRequest) {
        // Check if username already exists
        if (userRepository.existsByUsername(registerRequest.getUsername())) {
            log.warn("Registration failed: Username '{}' already exists", registerRequest.getUsername());
            throw new IllegalArgumentException("Username already exists");
        }

        // Check if email already exists
        if (userRepository.existsByEmail(registerRequest.getEmail())) {
            log.warn("Registration failed: Email '{}' already exists", registerRequest.getEmail());
            throw new IllegalArgumentException("Email already exists");
        }

        // Create new user
        User user = User.builder()
                .username(registerRequest.getUsername())
                .email(registerRequest.getEmail())
                .password(passwordEncoder.encode(registerRequest.getPassword()))
                .role(Role.USER) // Default role is USER
                .build();

        // Save user to database
        user = userRepository.save(user);

        log.info("New user registered successfully: {}", user.getUsername());

        // Automatically log in the user
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                        registerRequest.getUsername(),
                        registerRequest.getPassword()
                )
        );
        SecurityContextHolder.getContext().setAuthentication(authentication);

        // Build and return response
        return LoginResponse.builder()
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .role(user.getRole())
                .message("Registration successful")
                .build();
    }
}

File: frontend/src/services/authService.js
import api from './api';

/**
 * Authentication service for handling login/logout operations.
 */
export const authService = {
  /**
   * Login with username and password.
   * @param {string} username - User's username
   * @param {string} password - User's password
   * @returns {Promise} User data on successful login
   */
  async login(username, password) {
    const response = await api.post('/auth/login', { username, password });
    if (response.data) {
      // Store user info in session storage
      sessionStorage.setItem('user', JSON.stringify(response.data));
    }
    return response.data;
  },

  /**
   * Register a new user.
   * @param {string} username - User's username
   * @param {string} email - User's email
   * @param {string} password - User's password
   * @returns {Promise} User data on successful registration
   */
  async register(username, email, password) {
    const response = await api.post('/auth/register', { username, email, password });
    if (response.data) {
      // Store user info in session storage (auto-login after registration)
      sessionStorage.setItem('user', JSON.stringify(response.data));
    }
    return response.data;
  },

  /**
   * Logout current user.
   */
  async logout() {
    try {
      await api.post('/auth/logout');
    } catch (error) {
      console.error('Logout error:', error);
    } finally {
      // Clear session storage
      sessionStorage.removeItem('user');
    }
  },

  /**
   * Get current user from session storage.
   * @returns {Object|null} User object or null if not authenticated
   */
  getCurrentUser() {
    const userStr = sessionStorage.getItem('user');
    return userStr ? JSON.parse(userStr) : null;
  },

  /**
   * Check if user is authenticated.
   * @returns {boolean} True if user is logged in
   */
  isAuthenticated() {
    return !!this.getCurrentUser();
  }
};
